### PR TITLE
Enable idle rate counter on demand

### PR DESCRIPTION
**Pull request by hkaiser**
This removes the need for compile time configuration of the idle-rate counters. Those will be enabled by default, but activated only when needed.

### CODE REVIEW

File: hpx/util/coroutine/detail/context_windows_fibers.hpp [hunk 1]
```diff
@@ -49,7 +51,7 @@
 #include <boost/atomic.hpp>
 
 #if defined(HPX_HAVE_SWAP_CONTEXT_EMULATION)
-extern "C" void switch_to_fiber(void* lpFiber) throw();
+extern "C" void HPX_EXPORT switch_to_fiber(void* lpFiber) throw();
```
Comment (K-ballo) [inline hpx/util/coroutine/detail/context_windows_fibers.hpp]: Is this change related to the PR? If so, how?

Comment (hkaiser) [inline hpx/util/coroutine/detail/context_windows_fibers.hpp]: Probably not, but for some weird reason I was not able to make it work without this change :/ It should probably a separate bug-fix, but I think it does not harm if it is left as a part of this PR.

**Comment by hkaiser** [comment]
Are there any objections to merge this?

**Comment by K-ballo** [comment]
I would have liked to see a proper review here, as this was already changed in the opposite direction (about a year or two ago), and concerns about it were raised on IRC. I am not familiar with the implementation of counters nor with the rationale for disabling idle-rate (overheads?) to do that review, but the general code looks ok and the pattern seems to be already used for other counters (incidentally, access to those `extern bools` are theoretically races).

This is not an objection.

**Comment by hkaiser** [comment]
> I would have liked to see a proper review here, as this was already changed in the opposite direction (about a year or two ago), and concerns about it were raised on IRC

No, it never was implemented this way, it was always a compile time switch. @pagrubel was doing some analysis a while back. She said that it had no noticeable effect on performance. But ok, let's do a full review here.

> incidentally, access to those  extern bools  are theoretically races

Yah, it's what's called a benign data race.

**Comment by K-ballo** [comment]
> I would have liked to see a proper review here, as this was already changed in the opposite direction (about a year or two ago), and concerns about it were raised on IRC
> 
> No, it never was implemented this way, it was always a compile time switch. @pagrubel was doing some analysis a while back. She said that it had no noticeable effect on performance. But ok, let's do a full review here.

I'm not asking for a full review. The code looks sane, and anyone interested had plenty enough time to be made aware of the proposed change. I would have just liked the PR to be signed off by those who have expressed concern.

>     incidentally, access to those extern bools are theoretically races
> 
> Yah, it's what's called a benign data race.

They actually look like undefined behavior, access to them is still potentially concurrent regardless of the architecture and whatever codegen they produce. I'm not asking for this to change either, I know it to work and be portable (for now) despite not being standard conformant.

File: hpx/runtime/threads/policies/thread_queue.hpp [hunk 2]
```diff
@@ -60,6 +60,10 @@ namespace hpx { namespace threads { namespace policies
     // startup code
     extern bool minimal_deadlock_detection;
 #endif
+    // We control whether to collect idle rates using this global bool.
+    // It will be set by any of the related performance counters. Once set it
+    // stays set, thus no race conditions will occur.
+    extern bool maintain_idle_rates;
```
Comment (sithhell) [inline hpx/runtime/threads/policies/thread_queue.hpp]: Why does the bool have to be defined here as well? Especially since it is not used in that header anywhere?

**Comment by sithhell** [comment]
I was hoping to see some comments from @pagrubel regarding the performance impact (if any) of those changes.

**Comment by pagrubel** [comment]
Yes, I’d like to run some tests. I haven’t had time as of yet. Now, I see hermione is having problems. Can you wait a bit?

Pat
On May 7, 2015, at 7:07 PM, Hartmut Kaiser <notifications@github.com<mailto:notifications@github.com>> wrote:


Are there any objections to merge this?

—
Reply to this email directly or view it on GitHub<https://github.com/STEllAR-GROUP/hpx/pull/1489#issuecomment-100058386>.

Patricia A. Grubel
Ph.D Candidate
Klipsch School of Electrical and Computer Engineering
New Mexico State University
575-434-2538 (home/work)
575-491-8206 (cell)

File: docs/manual/existing_performance_counters.qbk [hunk 3]
```diff
@@ -551,8 +551,8 @@ system and application performance.
          is `worker-thread#*` the counter will return the average time spent
          executing one __hpx__-thread for all worker threads separately.
          This counter is available only if the configuration time constants
-         `HPX_THREAD_MAINTAIN_CUMULATIVE_COUNTS` (default: ON) and
-         `HPX_THREAD_MAINTAIN_IDLE_RATES` are set to `ON` (default: OFF).]
```
Comment (K-ballo) [inline docs/manual/existing_performance_counters.qbk]: These names look old, shouldn't these have been `HPX_WITH_...` already?

Comment (hkaiser) [inline docs/manual/existing_performance_counters.qbk]: Indeed, will fix

Comment (hkaiser) [inline docs/manual/existing_performance_counters.qbk]: OTOH, this branch might be old enough to still have the old names.

Comment (K-ballo) [inline docs/manual/existing_performance_counters.qbk]: The new names are used in the `cmake_variables` file above

**Comment by hkaiser** [comment]
Note to self: make sure this counter interacts with the existence of RTDSC or similar.

**Comment by sithhell** [comment]
@pagrubel do you still need to run your performance regression tests? Could you share your scripts so that someone could take over that duty if you lack the time at the moment?

**Comment by pagrubel** [comment]
I ran the data and will look at it by tomorrow

Pat
On May 31, 2015, at 2:29 PM, Thomas Heller <notifications@github.com<mailto:notifications@github.com>> wrote:


@pagrubel<https://github.com/pagrubel> do you still need to run your performance regression tests? Could you share your scripts so that someone could take over that duty if you lack the time at the moment?

**Comment by pagrubel** [comment]
Doesn't lgtm
I'm seeing some really bad idle_rate counts, bad variances between samples for some core counts.

**Comment by hkaiser** [comment]
Ok, thanks Pat! I'll close this without merging, then.
